// firebase-tools/lib/emulator/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url = require("url");
const DEFAULT_PORTS = {
    ui: 4000,
    hub: 4400,
    logging: 4500,
    hosting: 5000,
    functions: 5001,
    firestore: 8080,
    database: 9000,
    pubsub: 8085,
};
exports.FIND_AVAILBLE_PORT_BY_DEFAULT = {
    ui: true,
    hub: true,
    logging: true,
    hosting: false,
    functions: false,
    firestore: false,
    database: false,
    pubsub: false,
};
exports.EMULATOR_DESCRIPTION = {
    ui: "Emulator UI",
    hub: "emulator hub",
    logging: "Logging Emulator",
    hosting: "Hosting Emulator",
    functions: "Functions Emulator",
    firestore: "Firestore Emulator",
    database: "Database Emulator",
    pubsub: "Pub/Sub Emulator",
};
const DEFAULT_HOST = "localhost";
class Constants {
    static getServiceName(service) {
        switch (service) {
            case this.SERVICE_FIRESTORE:
                return "firestore";
            case this.SERVICE_REALTIME_DATABASE:
                return "database";
            case this.SERVICE_PUBSUB:
                return "pubsub";
            case this.SERVICE_ANALYTICS:
                return "analytics";
            case this.SERVICE_AUTH:
                return "auth";
            case this.SERVICE_CRASHLYTICS:
                return "crashlytics";
            case this.SERVICE_REMOTE_CONFIG:
                return "remote config";
            case this.SERVICE_STORAGE:
                return "storage";
            case this.SERVICE_TEST_LAB:
                return "test lab";
            default:
                return service;
        }
    }
    static getDefaultHost(emulator) {
        return DEFAULT_HOST;
    }
    static getDefaultPort(emulator) {
        return DEFAULT_PORTS[emulator];
    }
    static getHostKey(emulator) {
        return `emulators.${emulator.toString()}.host`;
    }
    static getPortKey(emulator) {
        return `emulators.${emulator.toString()}.port`;
    }
    static description(name) {
        return exports.EMULATOR_DESCRIPTION[name];
    }
    static normalizeHost(host) {
        let normalized = host;
        if (!normalized.startsWith("http")) {
            normalized = `http://${normalized}`;
        }
        const u = url.parse(normalized);
        return u.hostname || DEFAULT_HOST;
    }
}
exports.Constants = Constants;
Constants.DEFAULT_DATABASE_EMULATOR_NAMESPACE = "fake-server";
Constants.FIRESTORE_EMULATOR_HOST = "FIRESTORE_EMULATOR_HOST";
Constants.FIREBASE_DATABASE_EMULATOR_HOST = "FIREBASE_DATABASE_EMULATOR_HOST";
Constants.SERVICE_FIRESTORE = "firestore.googleapis.com";
Constants.SERVICE_REALTIME_DATABASE = "firebaseio.com";
Constants.SERVICE_PUBSUB = "pubsub.googleapis.com";
Constants.SERVICE_ANALYTICS = "app-measurement.com";
Constants.SERVICE_AUTH = "firebaseauth.googleapis.com";
Constants.SERVICE_CRASHLYTICS = "fabric.io";
Constants.SERVICE_REMOTE_CONFIG = "firebaseremoteconfig.googleapis.com";
Constants.SERVICE_STORAGE = "storage.googleapis.com";
Constants.SERVICE_TEST_LAB = "testing.googleapis.com";
