// firebase-tools/lib/emulator/loggingEmulator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
const constants_1 = require("./constants");
const triple_beam_1 = require("triple-beam");
const WebSocket = require("ws");
const TransportStream = require("winston-transport");
const logger = require("../logger");
const ansiStrip = require("cli-color/strip");
class LoggingEmulator {
    constructor(args) {
        this.args = args;
    }
    start() {
        this.transport = new WebSocketTransport();
        this.transport.start(this.getInfo());
        logger.add(this.transport);
        return Promise.resolve();
    }
    connect() {
        return Promise.resolve();
    }
    stop() {
        var _a;
        logger.remove(this.transport);
        if ((_a = this.transport) === null || _a === void 0 ? void 0 : _a.wss) {
            const wss = this.transport.wss;
            return new Promise((resolve, reject) => {
                wss.close((err) => {
                    if (err)
                        reject(err);
                    else
                        resolve();
                });
            });
        }
        return Promise.resolve();
    }
    getInfo() {
        const host = this.args.host || constants_1.Constants.getDefaultHost(types_1.Emulators.LOGGING);
        const port = this.args.port || constants_1.Constants.getDefaultPort(types_1.Emulators.LOGGING);
        return {
            host,
            port,
        };
    }
    getName() {
        return types_1.Emulators.LOGGING;
    }
}
exports.LoggingEmulator = LoggingEmulator;
LoggingEmulator.LOGGING_EMULATOR_ENV = "FIREBASE_LOGGING_EMULATOR_HOST";
class WebSocketTransport extends TransportStream {
    constructor(options = {}) {
        super(options);
        this.connections = [];
        this.history = [];
        this.setMaxListeners(30);
    }
    start(options) {
        this.wss = new WebSocket.Server(options);
        this.wss.on("connection", (ws) => {
            this.connections.push(ws);
            this.history.forEach((bundle) => {
                ws.send(JSON.stringify(bundle));
            });
        });
    }
    log(info, next) {
        setImmediate(() => this.emit("logged", info));
        const bundle = {
            level: info.level,
            data: {},
            timestamp: new Date().getTime(),
            message: "",
        };
        const splat = [info.message, ...(info[triple_beam_1.SPLAT] || [])]
            .map((value) => {
            if (typeof value == "string") {
                try {
                    bundle.data = Object.assign(Object.assign({}, bundle.data), JSON.parse(value));
                    return null;
                }
                catch (err) {
                    return value;
                }
            }
            else {
                bundle.data = Object.assign(Object.assign({}, bundle.data), value);
            }
        })
            .filter((v) => v);
        bundle.message = splat.join(" ");
        if (bundle.data && bundle.data.metadata && bundle.data.metadata.level) {
            bundle.level = bundle.data.metadata.level.toLowerCase();
        }
        else {
            bundle.level = bundle.level.toLowerCase();
        }
        if (bundle.data && bundle.data.metadata && bundle.data.metadata.message) {
            bundle.message = bundle.data.metadata.message;
        }
        bundle.message = ansiStrip(bundle.message);
        this.history.push(bundle);
        this.connections.forEach((ws) => {
            ws.send(JSON.stringify(bundle));
        });
        if (next) {
            next();
        }
    }
}
