// firebase-tools/lib/emulator/ui.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
const downloadableEmulators = require("./downloadableEmulators");
const registry_1 = require("./registry");
const hub_1 = require("./hub");
const error_1 = require("../error");
const constants_1 = require("./constants");
class EmulatorUI {
    constructor(args) {
        this.args = args;
    }
    start() {
        if (!registry_1.EmulatorRegistry.isRunning(types_1.Emulators.HUB)) {
            throw new error_1.FirebaseError(`Cannot start ${constants_1.Constants.description(types_1.Emulators.UI)} without ${constants_1.Constants.description(types_1.Emulators.HUB)}!`);
        }
        const hubInfo = registry_1.EmulatorRegistry.get(types_1.Emulators.HUB).getInfo();
        const { auto_download, host, port, projectId } = this.args;
        const env = {
            HOST: host.toString(),
            PORT: port.toString(),
            GCLOUD_PROJECT: projectId,
            [hub_1.EmulatorHub.EMULATOR_HUB_ENV]: `${hubInfo.host}:${hubInfo.port}`,
        };
        return downloadableEmulators.start(types_1.Emulators.UI, { auto_download }, env);
    }
    connect() {
        return Promise.resolve();
    }
    stop() {
        return downloadableEmulators.stop(types_1.Emulators.UI);
    }
    getInfo() {
        return {
            host: this.args.host,
            port: this.args.port,
        };
    }
    getName() {
        return types_1.Emulators.UI;
    }
}
exports.EmulatorUI = EmulatorUI;
